### PR TITLE
Serialize/Deserialize multishape

### PR DESCRIPTION
**Pull request by caillef**

### CODE REVIEW

File: core/serialization_v6.c [hunk 1]
```diff
@@ -705,6 +830,22 @@ Shape *serialization_v6_load_shape(Stream *s,
...
                 break;
             }
+            case P3S_CHUNK_ID_NB_SHAPES: {
+                sizeRead = chunk_v6_read_nb_shapes(s, &nbShapes);
+                if (sizeRead == 0) {
+                    cclog_error("error while reading palette");
+                    error = true;
+                    break;
+                }
+                free(shapes);
```
Comment (ArthurVoxowl) [inline core/serialization_v6.c:840]: Remove this (see previous comment)

File: core/serialization_v6.c [hunk 2]
```diff
@@ -705,6 +830,22 @@ Shape *serialization_v6_load_shape(Stream *s,
...
+                shapes = malloc(nbShapes * sizeof(*shapes));
+                memset(shapes, 0, nbShapes * sizeof(*shapes));
```
Comment (ArthurVoxowl) [inline core/serialization_v6.c:842]: Please use `sizeof(Shape*)` for clarity, this line and the line above

File: core/serialization_v6.c [hunk 3]
```diff
@@ -973,6 +1153,27 @@ uint32_t chunk_v6_read_shape(Stream *s,
         cursor = (void *)((uint8_t *)cursor + 1);
         totalSizeRead += 1; // size of chunk id
         switch (chunkID) {
+            case P3S_CHUNK_ID_SHAPE_ID: {
+                memcpy(shapeId, cursor, sizeof(*shapeId));
```
Comment (ArthurVoxowl) [inline core/serialization_v6.c:1157]: Use `sizeof(uint32_t)` for clarity, the way we advance cursor is also based on the type. Overall, please use `sizeof(type)` in every similar line below

File: core/serialization_v6.c [hunk 4]
```diff
@@ -1718,6 +1947,41 @@ bool chunk_v6_shape_create_and_write_compressed_buffer(const Shape *shape,
     return true;
 }
 
+void _chunk_v6_nb_shapes_create_and_write_uncompressed_buffer(uint32_t nbShapes,
+                                                              uint32_t *uncompressedSize,
+                                                              void **uncompressedData) {
+    *uncompressedSize = sizeof(nbShapes);
```
Comment (ArthurVoxowl) [inline core/serialization_v6.c:1953]: Use `sizeof(uint32_t)`

File: core/serialization_v6.c [hunk 5]
```diff
@@ -541,40 +657,44 @@ bool serialization_v6_save_shape_as_buffer(const Shape *shape,
         ok = write_preview_chunk_in_buffer(buf + cursor, previewData, previewDataSize, &cursor);
```
Comment (ArthurVoxowl) [inline core/serialization_v6.c:657]: Before writing preview here, you should be writing `nbShapesCompressedDataSize` into destination buffer `buf` using eg. `write_compressed_chunk_in_buffer` or a similar function

**Review by ArthurVoxowl — CHANGES_REQUESTED** [review]
There is an issue in `serialization_v6_save_shape_as_buffer` as well as a number of minor syntax tweaks, please check the comments!

You also need to add a `case P3S_CHUNK_ID_NB_SHAPES:` in the switch in  `serialization_v6_get_preview_data`, so it can skip the chunk correctly when looking for the preview

File: cubzh-file-format-3zh.txt [hunk 6]
```diff
@@ -0,0 +1,216 @@
+Cubzh .3zh File Format [10/20/2022]
+
+File
+{
+    Header
+
+    // PNG Preview
+    Chunk 'PREVIEW' : optional
+
+    Chunk 'NB_SHAPES' : N, optional
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:10]: We don't need this one. We'll just count shapes while processing file.

File: cubzh-file-format-3zh.txt [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+-------------------------------------------------------------------------------
+
+
+1. Header
+-------------------------------------------------------------------------------
+# Bytes  | Type       | Value
+-------------------------------------------------------------------------------
+6        | byte       | magic bytes 'CUBZH!' : 'C' 'U' 'B' 'Z' 'H' '!', 'C' is first
+4        | int        | version number : 6
+1        | byte       | p3s compression method : 0 (none), 1 (zip), 2 (count)
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:31]: `count` is not a compression method, it just indicated the number of supported compression methods in the enum. We only support 0 (none) or 1 (zip) for now.

File: cubzh-file-format-3zh.txt [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+4        | uint       | total size
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:32]: Maybe we should precise that this is the total size of data(compressed or not) coming after this number. (but the way, it's a `uint32_t`)

File: cubzh-file-format-3zh.txt [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+    // POIs rotation
+    SubChunk 'SHAPE_POINT_ROTATION' : optional
+
+    ...
+
+    SubChunk 'SHAPE_POINT_ROTATION' : optional
+
+    // Baked lighting
+    SubChunk 'SHAPE_BAKED_LIGHTING' : optional
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:114]: `SHAPE_PALETTE` SubChunk missing.

File: cubzh-file-format-3zh.txt [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+4        | int        | version number : 6
+1        | byte       | p3s compression method : 0 (none), 1 (zip), 2 (count)
+4        | uint       | total size
+
+
+2. Chunk Structure (not SubChunks)
+-------------------------------------------------------------------------------
+# Bytes  | Type       | Value
+-------------------------------------------------------------------------------
+1        | byte       | chunk id
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:39]: byte -> uint8

File: cubzh-file-format-3zh.txt [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+4        | int        | num bytes of chunk content (N)
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:40]: int -> uint32

File: cubzh-file-format-3zh.txt [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+1        | byte       | is compressed
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:41]: byte -> uint8 (0: not compressed - 1: compressed)

File: cubzh-file-format-3zh.txt [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+4        | int        | num bytes of chunk content uncompressed (M)
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:42]: int -> uint32

File: cubzh-file-format-3zh.txt [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+4        | uint       | number of objects
+-------------------------------------------------------------------------------
+
+
+5. Chunk id 'PALETTE' : optional, contains colors. Not related to any shapes in the file (except legacy single shape files, this chunk is the palette of the shape)
+-------------------------------------------------------------------------------
+# Bytes  | Type       | Value
+-------------------------------------------------------------------------------
+1        | byte       | color count (N)
+4 x N    | int        | (r, g, b, alpha) : 1 byte for each entry
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:68]: int -> uint8 x 4
not 4 x N (would be 4 times too much) - each color is 4 bytes

File: cubzh-file-format-3zh.txt [hunk 15]
```diff
@@ -0,0 +1,216 @@
...
+1 x N    | bool       | emissive flag
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:69]: let's use uint8s instead of bools

File: cubzh-file-format-3zh.txt [hunk 16]
```diff
@@ -0,0 +1,216 @@
...
+    // PNG Preview
+    Chunk 'PREVIEW' : optional
+
+    Chunk 'NB_SHAPES' : N, optional
+
+    // Palette
+    Chunk 'PALETTE' : optional
+
+    // Objects (x N)
+    Chunk 'SHAPE' / Chunk 'OBJECT'
```
Comment (aduermael) [inline cubzh-file-format-3zh.txt:16]: SHAPE is also optional. 
A file could only contain a PALETTE

**Review by aduermael — COMMENTED** [review]
A few comments on the file format.

File: core/serialization.c [hunk 17]
```diff
@@ -115,15 +135,21 @@ Shape *serialization_load_shape(Stream *s,
...
+    if (doubly_linked_list_node_count(list) == 0) {
+        doubly_linked_list_free(list);
+        list = NULL;
+        cclog_error("[serialization_load_resources] no resources found");
+    }
+    
+    // set fullname if containing a root shape
+    Shape *shape = resources_get_root_shape(list);
     if (shape != NULL) {
-        shape_shrink_box(shape);
```
Comment (ArthurVoxowl) [inline core/serialization.c:119]: You added this call inside `serialization_load_resources_v6`, but it is now missing for a v5 shape. You could add it in `serialization_v5_load_shape`

File: core/serialization.c [hunk 18]
```diff
@@ -115,15 +135,21 @@ Shape *serialization_load_shape(Stream *s,
     stream_free(s);
 
     // shrink box once all blocks were added to update box origin
```
Comment (ArthurVoxowl) [inline core/serialization.c:137]: Please remove this comment if the associated code was removed (or move it where the new calls are done)

**Review by ArthurVoxowl — CHANGES_REQUESTED** [review]

**Review by aduermael — APPROVED** [review]
Looks good to me!
Thanks @caillef!

**Review by ArthurVoxowl — APPROVED** [review]

**Pull request merged**